### PR TITLE
Reporting errors on foreach ref/readonly declarations in async/iterator methods

### LINKED ISSUE #26585: "csc.exe" exited with code -2146232797
**Version Used**: 

Microsoft Visual Studio Professional 2017 Preview (2)
Version 15.7.0 Preview 5.0
VisualStudio.15.Preview/15.7.0-pre.5.0+27625.0
Microsoft .NET Framework
Version 4.7.02556

**Steps to Reproduce**:

Try to compile a foreach cycle with a ref iterator variable in an async/iterator method
e.g.

```cs
using System;
class E
{
  public class Enumerator
  {
    public ref readonly int Current => throw new NotImplementedException();
    public bool MoveNext() => throw new NotImplementedException();
  }
  public Enumerator GetEnumerator() => new Enumerator();
}
class C
{
  public async void M()
  {
    foreach (ref readonly int x in new E()) {}
  }
}
```

or 

```cs
using System;
class E
{
  public class Enumerator
  {
    public ref readonly int Current => throw new NotImplementedException();
    public bool MoveNext() => throw new NotImplementedException();
  }
  public Enumerator GetEnumerator() => new Enumerator();
}
class C
{
  public System.Collections.Generic.IEnumerable<int>  M()
  {
    yield return 0;
    foreach (ref readonly int x in new E()) {}
  }
}
```

**Expected Behavior**:

Meaningful error.

**Actual Behavior**:

```"csc.exe" exited with code -2146232797```

Stacktrace from [sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDZICYgNQA+AAgAwAERAjANwCwAUEQExkCiDA3g2RQMwUtWAOwCuYRBBgB7BNzJd6PHkX4I4AMzJqI2KUIwBPMmiEwyAYREI1psgF4AfGRgALBFIDuZIXC8A5KRgASTAABww4cVM4bFYADwBjOFCYND0ACgBKOkUlFTJgKSkMMgBZKQA3OD84OJgs+ydXdy8ff0CQ8Mi4aNjE5NSM7LkAXzl84TEJaQQyAHE4GEnxBEkZBsdvXzZRFbWELJyxxhZzTnH+IgAOCgAWMqy5BSUydRk4CASXMnS1TW1dPojCYzHFjEItl5WFlMvJjjxjiMgA==)
```
    System.InvalidCastException: Unable to cast object of type 'ForEachLocalSymbol' to type 'Microsoft.CodeAnalysis.CSharp.Symbols.SynthesizedLocal'.

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitAssignmentOperator(BoundAssignmentOperator node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 774

       at Microsoft.CodeAnalysis.CSharp.BoundAssignmentOperator.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 1245

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 97

       at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 225

       at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeVisitors.cs:line 206

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 82

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.AsyncMethodToStateMachineRewriter.VisitExpressionStatement(BoundExpressionStatement node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncMethodToStateMachineRewriter.cs:line 252

       at Microsoft.CodeAnalysis.CSharp.BoundExpressionStatement.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2693

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.VisitSequencePointWithSpan(BoundSequencePointWithSpan node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 9006

       at Microsoft.CodeAnalysis.CSharp.BoundSequencePointWithSpan.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2267

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 37

       at Microsoft.CodeAnalysis.CSharp.Symbols.MethodToClassRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\MethodToClassRewriter.cs:line 137

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.<>c__DisplayClass41_0.<VisitBlock>b__0() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.PossibleIteratorScope(ImmutableArray`1 locals, Func`1 wrapped) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 308

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2303

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 37

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.VisitStatementList(BoundStatementList node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 9266

       at Microsoft.CodeAnalysis.CSharp.BoundStatementList.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 4076

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 37

       at Microsoft.CodeAnalysis.CSharp.Symbols.MethodToClassRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\MethodToClassRewriter.cs:line 137

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.<>c__DisplayClass41_0.<VisitBlock>b__0() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.PossibleIteratorScope(ImmutableArray`1 locals, Func`1 wrapped) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 265

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2303

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitTryStatement(BoundTryStatement node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 801

       at Microsoft.CodeAnalysis.CSharp.BoundTryStatement.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 3489

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 37

       at Microsoft.CodeAnalysis.CSharp.Symbols.MethodToClassRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\MethodToClassRewriter.cs:line 137

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.<>c__DisplayClass41_0.<VisitBlock>b__0() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.PossibleIteratorScope(ImmutableArray`1 locals, Func`1 wrapped) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 308

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2303

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 37

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.VisitStatementList(BoundStatementList node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 9266

       at Microsoft.CodeAnalysis.CSharp.BoundStatementList.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 4076

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 37

       at Microsoft.CodeAnalysis.CSharp.Symbols.MethodToClassRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\MethodToClassRewriter.cs:line 137

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.<>c__DisplayClass41_0.<VisitBlock>b__0() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.PossibleIteratorScope(ImmutableArray`1 locals, Func`1 wrapped) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 265

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitBlock(BoundBlock node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 657

       at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs:line 2303

       at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\BoundTree\BoundTreeRewriter.cs:line 87

       at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\MethodToStateMachineRewriter.cs:line 649

       at Microsoft.CodeAnalysis.CSharp.AsyncMethodToStateMachineRewriter.GenerateMoveNext(BoundStatement body, MethodSymbol moveNextMethod) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncMethodToStateMachineRewriter.cs:line 120

       at Microsoft.CodeAnalysis.CSharp.AsyncRewriter.GenerateMoveNext(SynthesizedImplementationMethod moveNextMethod) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncRewriter.cs:line 248

       at Microsoft.CodeAnalysis.CSharp.AsyncRewriter.GenerateMethodImplementations() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncRewriter.cs:line 124

       at Microsoft.CodeAnalysis.CSharp.StateMachineRewriter.Rewrite() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\StateMachineRewriter\StateMachineRewriter.cs:line 106

       at Microsoft.CodeAnalysis.CSharp.AsyncRewriter.Rewrite(BoundStatement body, MethodSymbol method, Int32 methodOrdinal, VariableSlotAllocator slotAllocatorOpt, TypeCompilationState compilationState, DiagnosticBag diagnostics, AsyncStateMachine& stateMachineType) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Lowering\AsyncRewriter\AsyncRewriter.cs:line 68

       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, Boolean instrumentForDynamicAnalysis, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& dynamicAnalysisSpans, DiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 1326

       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 1079

       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 494

       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass21_0.<CompileNamedTypeAsTask>b__0() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 384

       at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\Core\Portable\InternalUtilities\UICultureUtilities.cs:line 136

       at System.Threading.Tasks.Task.InnerInvoke()

       at System.Threading.Tasks.Task.Execute()

    --- End of stack trace from previous location where exception was thrown ---

       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers() in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 299

       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean emitTestCoverageData, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 144

       at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, Boolean emitMetadataOnly, Boolean emitTestCoverageData, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 2533

       at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream metadataPEStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, CompilationTestData testData, CancellationToken cancellationToken) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\Core\Portable\Compilation\Compilation.cs:line 2195

       at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, Stream metadataPEStream, CancellationToken cancellationToken) in D:\$TC\work\c99972096bc51e45\!roslyn\sources\dotnet\src\Compilers\Core\Portable\Compilation\Compilation.cs:line 2143

       at SharpLab.Server.Compilation.Compiler.<TryCompileToStreamAsync>d__0.MoveNext()

    --- End of stack trace from previous location where exception was thrown ---

       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

       at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)

       at SharpLab.Server.MirrorSharp.SlowUpdate.<ProcessAsync>d__7.MoveNext()

    --- End of stack trace from previous location where exception was thrown ---

       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

       at MirrorSharp.Internal.Handlers.SlowUpdateHandler.<ExecuteAsync>d__4.MoveNext()

    --- End of stack trace from previous location where exception was thrown ---

       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

       at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)

       at MirrorSharp.Internal.Connection.<ReceiveAndProcessInternalAsync>d__13.MoveNext()

    --- End of stack trace from previous location where exception was thrown ---

       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)

       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

       at MirrorSharp.Internal.Connection.<ReceiveAndProcessAsync>d__12.MoveNext()
```

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #26585

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/ForEachTests.cs [hunk 1]
```diff
@@ -3196,5 +3196,137 @@ static void Main()
...
+    {
+        await Task.CompletedTask;
+
+        foreach (ref int x in new E())
+        {
+            System.Console.Write(x);
+        }
+    }
+}").VerifyDiagnostics(
+                // (20,26): error CS8177: Async methods cannot have by reference locals
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ForEachTests.cs]: >by reference [](start = 68, length = 12)

Nit: consider "by-reference" as we do in [CS8172](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDZICYgNQA+AAgAwAERAjANwCwAUKRZQKx2MDMFATGQMJkA3gzKiKXIgA4KANjIBZABQBKISLEa0AOxhkAnmQC8ZEuw0aEcAGZltugB5H9ZsQF8GroA)

**Comment by jcouv** [comment]
Got connectivity back. Approved to merge (as we discussed). Thanks

**Comment by jcouv** [comment]
Thanks

**Comment by OmarTawfik** [comment]
@dotnet-bot test this please

**Comment by OmarTawfik** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest
@dotnet-bot  test windows_debug_spanish_unit32_prtest

**Comment by OmarTawfik** [comment]
@dotnet-bot test windows_debug_spanish_unit32_prtest 
@dotnet-bot test windows_debug_vs-integration_prtest

**Comment by OmarTawfik** [comment]
@dotnet-bot test windows_debug_spanish_unit32_prtest

**Comment by OmarTawfik** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest

**Pull request merged**